internal/scheduler/upsert.go

```diff
@@ -0,0 +1,323 @@
+		dbMessageId := pulsarMessageIdFromPulsarRecord(sqlWriteMessageId)
+
+		// If the id loaded from the database is more recent than the one provided, abort the transaction.
+		// Since the data we're trying to write is stale.
+		if writeMessageId, ok := writeMessageIds[dbMessageId.PartitionIdx()]; ok {
+			isGreater, err := dbMessageId.Greater(writeMessageId)
+			if err != nil {
+				return err
+			}
+			if isGreater {
+				staleWrites = append(staleWrites, StaleWrite{
+					WriteMessageId: writeMessageId,
+					DbMessageId:    dbMessageId,
+				})
+			}
+		}
+	}
+	if len(staleWrites) > 0 {
+		return &ErrStaleWrite{
+			Topic:       topicName,
+			StaleWrites: staleWrites,
+		}
+	}
+
+	// Otherwise, we have more recent data than what is already stored in the db that we should write.
+	// Update the message id in postgres to reflect the data to be written.
+	//
+	// TODO: Add a call to insert these in a single call.
+	for partitionIdx, writeMessageId := range writeMessageIds {
+		err = queries.UpsertMessageId(ctx, UpsertMessageIdParams{
+			Topic:        fmt.Sprintf("%s-partition-%d", topicName, partitionIdx),
+			LedgerID:     writeMessageId.LedgerID(),
+			EntryID:      writeMessageId.EntryID(),
+			BatchIdx:     writeMessageId.BatchIdx(),
+			PartitionIdx: writeMessageId.PartitionIdx(),
+		})
+		if err != nil {
+			return errors.WithStack(err)
+		}
+	}
+	return nil
+}
+
+func CopyProtocolUpsert(ctx context.Context, tx pgx.Tx, tableName string, schema string, records []interface{}) error {
+
+	// Write records into postgres.
+	// First, create a temporary table for loading data in bulk using the copy protocol.
+	// The table is created with the provided schema.
+	tempTableName := "insert" // TODO: Is this name unique per transaction?
+	_, err := tx.Exec(ctx, fmt.Sprintf("CREATE TEMPORARY TABLE %s %s ON COMMIT DROP;", tempTableName, schema))
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// Use the postgres-specific COPY wire protocol to load data into the new table in a single operation.
+	// The COPY protocol can be faster than repeated inserts for as little as 5 rows; see
+	// https://www.postgresql.org/docs/current/populate.html
+	// https://pkg.go.dev/github.com/jackc/pgx/v4#hdr-Copy_Protocol
+	//
+	// We're guaranteed there is at least one record.
+	names, _ := NamesValuesFromRecord(records[0])
+	if len(names) < 2 {
+		return errors.Errorf("Names() must return at least 2 elements, but got %v", names)
+	}
+	n, err := tx.CopyFrom(ctx,
+		pgx.Identifier{tempTableName},
+		names,
+		pgx.CopyFromSlice(len(records), func(i int) ([]interface{}, error) {
+			// TODO: Are we guaranteed that values always come in the order listed in the record? Otherwise we need to control the order.
+			_, values := NamesValuesFromRecord(records[i])
+			return values, nil
+		}),
+	)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	if n != int64(len(records)) {
```

Comment:
Is this worth checking?  It's not clear to me how this could ever occur given the guarantees given by the db?

Reply:
I think we may as well keep it. It doesn't cost us anything.